horovod/run/run_task.py

```diff
@@ -0,0 +1,36 @@
+# Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+import cloudpickle
+import sys
+from horovod.run.http.http_client import read_data_from_kvstore, put_data_into_kvstore
+
+
+def main(addr, port):
+    pickled_func = read_data_from_kvstore(addr, port, 'runfunc', 'func')
+    func = cloudpickle.loads(pickled_func)
+    try:
+        ret_val = func()
+    except BaseException as e:
+        sys.stderr.write("User function raise error: {error}".format(error=str(e)))
+        raise e
+    if ret_val:
```

Comment:
There are a few issues here we should think about.

This will resolve to `False` for valid return values like `0`.  You could check `ret_val is None` instead, but in some cases `None` may also be a valid return value.  It's also possible that multiple ranks can return something, in which case this approach will result in all but one of them being thrown out.

What I suggest we do instead is to fetch the rank from the environment (you can see an example of how to do this in `test_tensorflow.py::test_horovod_rank`).  Then we can have every rank publish its `ret_val` (regardless of whether it is `None` or not) to the KV store with a key unique to its rank.  Back in driver, we can then iterate through the ranks and fetch every return value, adding them to a list we return to the caller.

What do you think?

Reply:
Yes, I considered this way. But there is an issue: the `rank()` method should import `horovod.tensorflow` or ``horovod.pytorch` or some other backends. Here I don't know which backend user use. How to address this ? @tgaddair

Reply:
You can pull the rank directly from environment variables without needing to import a framework.  I would suggest adding this method to `horovod/run/common/util/env.py`:

```
def get_env_rank_and_size():
    rank_env = ['HOROVOD_RANK', 'OMPI_COMM_WORLD_RANK', 'PMI_RANK']
    size_env = ['HOROVOD_SIZE', 'OMPI_COMM_WORLD_SIZE', 'PMI_SIZE']

    for rank_var, size_var in zip(rank_env, size_env):
        rank = os.environ.get(rank_var)
        size = os.environ.get(size_var)
        if rank is not None and size is not None:
            return int(rank), int(size)

    # Default to rank zero and size one if there are no environment variables
    return 0, 1
```

You can use that to get the rank and size directly from the environment.

Reply:
@tgaddair Spectrum MPI was added yesterday #1441
I doubt whether this version MPI can support OMPI_COMM_WORLD_RANK.

Reply:
According to the [Spectrum MPI user guide](https://hpc.llnl.gov/sites/default/files/MPI-SpectrumUserGuide.pdf), you can access the rank using `OMPI_COMM_WORLD_RANK`, same as for Open MPI.  

@nvcastet can you confirm that this is correct (and that `OMPI_COMM_WORLD_SIZE` is the correct env variable for world size)?

Reply:
@tgaddair I confirm that `OMPI_COMM_WORLD_SIZE`, `OMPI_COMM_WORLD_LOCAL_SIZE `, `OMPI_COMM_WORLD_RANK `, and `OMPI_COMM_WORLD_LOCAL_RANK ` work with SMPI.